proposals/function-pointers.md

```diff
@@ -0,0 +1,245 @@
+# Function Pointers
+
+## Summary
+This proposal provides language constructs that expose IL opcodes that cannot currently be accessed efficiently,
+or at all, in C# today: `ldftn` and `calli`. These IL opcodes can be important in high performance code and developers
+need an effecient way to access them.
+
+## Motivation
+The motivations and background for this feature are described in the following issue (as is a 
+potential implementation of the feature): 
+
+https://github.com/dotnet/csharplang/issues/191
+
+This is an alternate design propsoal to [compiler intrinsics](https://github.com/dotnet/csharplang/blob/master/proposals/intrinsics.md)
+
+## Detailed Design 
+
+### funcptr
+The language will allow for the declaration of function pointers using the `funcptr` contextual keyword. The 
```

Comment:
Hmm, `funcptr`, kind of ugly for various reasons, including the fact that C#/.NET tends to deal in methods and not functions. And I'm no fan of the C/C++ syntax for function pointers but it's weird that C# data pointers are declared like C/C++ data pointers while C# function pointers, named or unnamed, require this distinct syntax.

And coupled with the names/no names issue, it's even more questionable. This introduces a very specific way to assign a name to an otherwise unnamed type even though such a feature (typedef basically) is requested from time to time. Not saying that C# should get typedef but this seems like a "let's rush things up and get this done" kind of thing.

Maybe `funcptr`, if added, should be just syntactic sugar for:
```
unsafe struct Action { 
    void (*_ptr)(); 
    Action(void (*ptr)()) => _ptr = ptr; 
    public void Invoke() => _ptr(); 
}
```
I guess this kind of shots down half of the proposal. But hey, just my 2 cents :)

Reply:
>Maybe funcptr, if added, should be just syntactic sugar for

In this case it can't really be. The function pointer concept is actually a CLI primitive value. The goal is to just directly expose that.

Reply:
> In this case it can't really be. The function pointer concept is actually a CLI primitive value. The goal is to just directly expose that.

Yep, I wasn't very clear. The idea was to expose that primitive value through a more primitive syntax - `void (*_ptr)();` or something similar, that doesn't need a keyword that tries to look like a delegate and ends up looking like a typedef.

Reply:
Maybe we can reduce the number of keywords by using the calling convention enum to describe the unmanaged calling convention details? And use the variant without parameter to map to the default calling convention (if we do something about it)?

`funcptr unmanaged(CallingConvention.Cdecl) int F1()`
`funcptr unmanaged int F2()`

Note that there may be more unmanaged calling conventions coming in future: Windows have introduced vectorcall calling convention some time ago that .NET does not support yet.